### PR TITLE
Revisit contract wide Wyvern proxy pre-approval.

### PR DESCRIPTION
**Pull request by cxkoda**
# Motivation
Wyvern proxies are currently pre-approved by hard-coding them into the ERC721 base contract, which means that their access cannot be revoked by users. Since the corresponding `ApprovalForAll` events are not emitted, most external indexing tools will not be able to correctly recognized the approval state for a given user. This can lead to situations where users have wrong information on the security of their tokens and cannot change it if desired.

# Implementation
Introducing a new wrapper class for pre-approvals that

- allows opt-outs on a per-wallet basis
- emits events during transfers

### CODE REVIEW

File: contracts/erc721/ERC721PreApproval.sol [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        if (super.isApprovedForAll(owner, operator)) {
+            return true;
+        }
+
+        if (state[owner] == State.Active) {
+            return OpenSeaGasFreeListing.isApprovedForAll(owner, operator);
+        }
+
+        return false;
```
Comment (aschlosberg) [inline contracts/erc721/ERC721PreApproval.sol:47]: Why not just `return state[owner] == State.Active && OpenSeaGas…`?

File: contracts/erc721/ERC721PreApproval.sol [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        if (operator == OpenSeaGasFreeListing.proxyFor(owner)) {
+            state[owner] = approved ? State.Active : State.Inactive;
+            emit ApprovalForAll(owner, operator, approved);
+        } else {
+            super._setApprovalForAll(owner, operator, approved);
+        }
+    }
+
+    /// @dev Checks if the receiver has an existing proxy. If not, the
+    /// pre-approval is disabled.
```
Comment (aschlosberg) [inline contracts/erc721/ERC721PreApproval.sol:67]: Nice. I think this is a very sensible default.

File: contracts/erc721/ERC721PreApproval.sol [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+/// @dev This wrapper allows users to revoke the pre-approval of their
+/// associated proxy and emits the corresponding events. This is necessary for
+/// external tools to index approvals correctly and inform the user.
+/// @dev The pre-approval is triggered on a per-wallet basis during the first
+/// transfer transactions. It will only be enabled for wallets with an existing
+/// proxy. Not having a proxy incurs a gas overhead.
+/// @dev This wrapper optimizes for the following scenario:
+/// - The majority of users already have a wyvern proxy
+/// - Most of them want to transfer tokens via wyvern exchanges
+abstract contract ERC721PreApproval is ERC721 {
```
Comment (aschlosberg) [inline contracts/erc721/ERC721PreApproval.sol:18]: If it's `is IERC721` then anyone can use this with arbitrary `ERC721` implementations, not just OZ.

File: contracts/erc721/ERC721PreApproval.sol [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    /// pre-approval is disabled.
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override {
+        super._beforeTokenTransfer(from, to, tokenId);
+
+        // Exclude burns
+        if (to != address(0) && state[to] == State.Active) {
```
Comment (aschlosberg) [inline contracts/erc721/ERC721PreApproval.sol:76]:
```
if (to == address(0) || state[to] != State.Active) {
  return
}
```

Rationale: https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html TL;DR it acts as a signal to the reader of the code that nothing more will happen in this branch.

File: tests/erc721/erc721_test.go [hunk 5]
```diff
@@ -162,6 +166,105 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			sim, nft := deploy(t)
+
+			if tt.hasProxy {
+				openseatest.SetProxyTB(t, sim, sim.Addr(tokenMinter), sim.Addr(proxy))
+			}
+
+			approvalEvents := make(chan *TestableERC721CommonEnumerableApprovalForAll)
```
Comment (aschlosberg) [inline tests/erc721/erc721_test.go:196]: It's much cleaner to use `FilterApprovalForAll()` instead of `WatchApprovalForAll()`. I think I may have introduced the channel + `Watch` pattern elsewhere, before I knew about `Filter`. This allows you to collect _all_ events into a `[]` slice before using `cmp.Diff`; the benefit is that it captures any extras without the `time.After()` hack and no need for the `expected` flag is you literally just test for a known (possibly empty) slice of events.

File: tests/erc721/erc721_test.go [hunk 6]
```diff
@@ -162,6 +166,105 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+			nft.TestableERC721CommonEnumerableFilterer.WatchApprovalForAll(&bind.WatchOpts{}, approvalEvents, nil, nil)
+			defer close(approvalEvents)
+
+			testEvent := func(expected bool, approved bool) func(*testing.T) {
+				return func(t *testing.T) {
+					select {
+					case got := <-approvalEvents:
+						got.Raw = types.Log{}
```
Comment (aschlosberg) [inline tests/erc721/erc721_test.go:204]: Use `cmpopts.IgnoreFields(TestableERC721CommonEnumerableApprovalForAll{}, "Raw")` and pass it to `ethtest.Comparers()` as an extra option.

File: tests/erc721/erc721_test.go [hunk 7]
```diff
@@ -162,6 +166,105 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
 
+func TestOpenSeaProxyPreApproval(t *testing.T) {
+
+	tests := []struct {
+		name            string
+		hasProxy        bool
+		wantPreApproved bool
+	}{
+		{
+			name:            "without proxy on mint",
```
Comment (aschlosberg) [inline tests/erc721/erc721_test.go:177]: "on transfer" perhaps? Because that's the general case, and we're using mint as an easy example of a transfer.

File: tests/erc721/erc721_test.go [hunk 8]
```diff
@@ -162,6 +166,105 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+					nft.SetApprovalForAll(sim.Acc(tokenMinter), sim.Addr(proxy), true),
+				)
+				got, err := nft.IsApprovedForAll(nil, sim.Addr(tokenMinter), sim.Addr(proxy))
+				want := true
+				if err != nil || got != want {
+					t.Errorf("IsApprovedForAll([owner], [proxy]) got %t, err = %v; want %t, nil err", got, err, want)
+				}
+
+				t.Run("ApprovalForAll event", testEvent(true, true))
+			})
```
Comment (aschlosberg) [inline tests/erc721/erc721_test.go:262]: These can be collapsed into a sub-table (the `tests := []struct` is called a [table-driven test](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)) because they differ only in the boolean.

**Review by aschlosberg — CHANGES_REQUESTED** [review]
This is a very elegant solution! Thank you

File: contracts/erc721/ERC721CommonEnumerable.sol [hunk 9]
```diff
@@ -32,6 +32,16 @@ contract ERC721CommonEnumerable is ERC721Common, ERC721Enumerable {
...
+    /// @dev Calls ERC721Common.setApprovalForAll to manage pre-approvals
+    /// related to OpenSea's gas-free listing functionality.
+    function setApprovalForAll(address operator, bool approved)
+        public
+        virtual
+        override(ERC721Common, ERC721, IERC721)
+    {
+        ERC721Common.setApprovalForAll(operator, approved);
+    }
+
```
[the remark this answers is not in the document]
Comment (cxkoda) [reply contracts/erc721/ERC721CommonEnumerable.sol:44]: Good question. To be honest, I added this to satisfy the CI pipeline. On my local machine this wasn't needed I I didn't fully understand why yet. Will investigate.

[the remark this answers is not in the document]
Comment (cxkoda) [reply contracts/erc721/ERC721CommonEnumerable.sol:44]: This is because `ERC721Enumerable` inherits from `ERC721`. Therefore, both `ERC721Enumerable` and `ERC721Common` have a `setApprovalForAll` method and you need to specify which one to use here.

File: contracts/erc721/ERC721PreApproval.sol [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+/// external tools to index approvals correctly and inform the user.
+/// @dev The pre-approval is triggered on a per-wallet basis during the first
+/// transfer transactions. It will only be enabled for wallets with an existing
+/// proxy. Not having a proxy incurs a gas overhead.
+/// @dev This wrapper optimizes for the following scenario:
+/// - The majority of users already have a wyvern proxy
+/// - Most of them want to transfer tokens via wyvern exchanges
+abstract contract ERC721PreApproval is ERC721 {
+    /// @dev It is important that Active remains at first position, since this
+    /// is the scenario that we are trying to optimize for.
```
[the remark this answers is not in the document]
Comment (cxkoda) [reply contracts/erc721/ERC721PreApproval.sol:20]: Agreed, will add a test.

Yes, this allows us to optimize for the `Active` case because no overhead is needed to store it.

File: contracts/erc721/ERC721PreApproval.sol (hunk 1 above)
Comment (cxkoda) [reply contracts/erc721/ERC721PreApproval.sol:47]: Afaik `OpenSeaGasFreeListing.isApprovedForAll` is called regardless of the outcome of `state[owner] == State.Active` in an `.. && ..` statement. So the idea was to save some gas for `Inactive`.

However, this might introduce a small overhead for `Active` (would have to measure it), which goes against the initial assumptions and optimization goals. Therefore, I will change it to a `.. && ..` pattern for now.

File: tests/erc721/erc721_test.go (hunk 7 above)
Comment (cxkoda) [reply tests/erc721/erc721_test.go:177]: I think there is a bit of ambiguity around the term. 
- In the general case, there's no assumptions on `from` and `to`. So they would also include burns (i.e. `to = 0`), for which we would expect a different behavior.
- In the more specific case, one would assume `from != to` with both being non-zero. So they wouldn't include mints.

Comment (cxkoda) [reply tests/erc721/erc721_test.go:177]: So I'd rather include two more tests for special transfers and burns.

File: contracts/erc721/ERC721PreApproval.sol (hunk 3 above)
Comment (cxkoda) [reply contracts/erc721/ERC721PreApproval.sol:18]: Not quite possible here, because I rely on OZ's callbacks like `_beforeTokenTransfer` and `balanceOf`. Maybe we could introduce a basic internal ERC721 layout (with custom callbacks) for us at some point?

File: contracts/erc721/ERC721PreApproval.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (cxkoda) [reply contracts/erc721/ERC721PreApproval.sol:20]: If I understood correctly, this is not testable from the outside because the state is never exported - so `abigen` will not translate it. Maybe we should effectively lock this by testing transaction costs?

**Comment by cxkoda** [comment]
Thank you for the review!
I tried to realize your points about the testing but I'm not quite sure if I was able to do so as intended. Could you maybe have another look?
Also testing for `State.Active == 0` was not possible because the struct is never exported, so `abigen` won't pick it up. I am a bit torn if I should refactor this into a boolean mapping and drop the enum completely, since we no only need a binary state. Any opinions?

File: contracts/erc721/ERC721PreApproval.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (aschlosberg) [reply contracts/erc721/ERC721PreApproval.sol:20]: On second thought, the default behaviour would break if `Active` was non-zero, correct? If so, I guess we can leave it.

If you think we should have an explicit test, I would add a `TestableERC721PreApproval.sol` with functions `ActiveState` and `InactiveState`

File: contracts/erc721/ERC721PreApproval.sol (hunk 1 above)
Comment (aschlosberg) [reply contracts/erc721/ERC721PreApproval.sol]: Solidity appears to have [short-circuit boolean logic](https://github.com/ethereum/solidity/issues/576).

File: tests/erc721/erc721_test.go [hunk 11]
```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
```
Comment (aschlosberg) [inline tests/erc721/erc721_test.go:329]: Shouldn't this be a no-op when they call `setApprovalForAll()` as their state is already active by default?

File: tests/erc721/erc721_test.go [hunk 12]
```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+			if tt.hasProxy {
+				openseatest.SetProxyTB(t, sim, sim.Addr(tokenReceiver), sim.Addr(proxy))
+			}
+
+			if tt.mints {
+				sim.Must(t, "Mint(%d)", notExists)(nft.Mint(sim.Acc(tokenReceiver), big.NewInt(notExists)))
+			}
+
+			sim.Must(t, "%T.setApprovalForAll(proxy,%v)", nft, tt.wantApproved)(
+				nft.SetApprovalForAll(sim.Acc(tokenReceiver), sim.Addr(proxy), tt.wantApproved),
```
Comment (aschlosberg) [inline tests/erc721/erc721_test.go:444]: Please use different fields for the action and for the expected result. Even if they're always the same value, it makes it much clearer what the intended actions will be when just scanning the table.

File: tests/erc721/erc721_test.go [hunk 13]
```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+			wantApproved: false,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: false,
+				},
+			},
+		},
+	}
```
Comment (aschlosberg) [inline tests/erc721/erc721_test.go:428]: I think we have a case that's impossible to catch, in which someone doesn't explicitly opt in but then later sets a proxy. Nothing we can really do about that, yeah?

File: contracts/erc721/ERC721PreApproval.sol (hunk 1 above)
Comment (cxkoda) [reply contracts/erc721/ERC721PreApproval.sol]: Oh cool, I was not aware of that! Thanks

File: tests/erc721/erc721_test.go [hunk 14]
```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
 	}
 }
 
+func testApprovalForAllEvents(t *testing.T, filter *ERC721Filterer, start uint64, want []*ERC721ApprovalForAll) {
```
[the remark this answers is not in the document]
Comment (cxkoda) [reply tests/erc721/erc721_test.go:174]: Makes total sense. Changed it

File: tests/erc721/erc721_test.go [hunk 15]
```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+			start := sim.BlockNumber().Uint64()
+
+			toAddr := common.HexToAddress("0x0")
+			if tt.to >= 0 {
+				toAddr = sim.Addr(tt.to)
+			}
+
+			if tt.hasProxy {
+				openseatest.SetProxyTB(t, sim, toAddr, sim.Addr(proxy))
+			}
```
[the remark this answers is not in the document]
Comment (cxkoda) [reply tests/erc721/erc721_test.go:282]: I left it intentionally outside because the table also allows to set `hasProxy: true` for burns. So to avoid unexpected behavior in case someone decides to do so.

File: tests/erc721/erc721_test.go [hunk 16]
```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+				if err != nil || got != tt.wantPreApproved {
+					t.Errorf("IsApprovedForAll([owner], [proxy]) got %t, err = %v; want %t, nil err", got, err, tt.wantPreApproved)
+				}
+			}
+		})
+	}
+}
+
+func TestOpenSeaProxyPreApprovalOptInOut(t *testing.T) {
+	sim, _, _ := deploy(t)
```
[the remark this answers is not in the document]
Comment (cxkoda) [reply tests/erc721/erc721_test.go:305]: This is actually necessary because we need to call `sim.Addr` to fill the expected events in the table.
I also didn't like adding it here but I couldn't find another solution within reasonable effort. Any idea how this could be avoided?

File: tests/erc721/erc721_test.go [hunk 17]
```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
...
+		gotEvents = append(gotEvents, iter.Event)
+	}
+
+	if diff := cmp.Diff(want, gotEvents, ignore...); diff != "" {
+		t.Errorf("After %T deployment and single ownership transfer; Transfer events diff (-want +got):\n%s", filter, diff)
+	}
+}
+
+func TestOpenSeaProxyPreApproval(t *testing.T) {
+	sim, _, _ := deploy(t)
```
[the remark this answers is not in the document]
Comment (cxkoda) [reply tests/erc721/erc721_test.go:198]: see below

File: tests/erc721/erc721_test.go (hunk 11 above)
Comment (cxkoda) [reply tests/erc721/erc721_test.go:329]: Here I tried to mimic the standard behavior of OZ's ERC721, which also emits multiple events if `setApprovalForAll` is called repeatedly.

File: tests/erc721/erc721_test.go (hunk 13 above)
Comment (cxkoda) [reply tests/erc721/erc721_test.go:428]: No we can absolute catch this by adding something like `hasProxyAfterMint` to `TestOpenSeaProxyPreApprovalOptInOut`

File: contracts/erc721/ERC721PreApproval.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (cxkoda) [reply contracts/erc721/ERC721PreApproval.sol:20]: Yes, it would break in that case. So I'm opting to not check this explicitly for now.

**Comment by cxkoda** [comment]
Thanks for the review! Implemented most of the comments. See a few remaining open ones above.

File: tests/erc721/erc721_test.go (hunk 13 above)
Comment (cxkoda) [reply tests/erc721/erc721_test.go:489]: oh I think I misunderstood earlier. I added an extra `doSetApproved` flag to the above that determines if the manual opt in is actually called. This way we can have something like
```go
{
	name:                 "creating proxy after minting without proxy",
	hasProxyBeforeMint:   false,
	mints:                true,
	createProxyAfterMint: true,
	doSetApproved:        false,
	wantApproved:         false,
	wantEvents:           nil,
},
```
event thought the table gets quite long now.

File: contracts/erc721/ERC721PreApproval.sol [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        // Exclude burns and inactive pre-approvals
+        if (to == address(0) || state[to] == State.Inactive) {
+            return;
+        }
+
+        address operator = OpenSeaGasFreeListing.proxyFor(to);
+
+        // Disable if `to` has no proxy
+        if (operator == address(0)) {
+            state[to] = State.Inactive;
```
Comment (aschlosberg) [inline contracts/erc721/ERC721PreApproval.sol:82]: Nice! This is the scenario I was worried about us not capturing in tests (owner gets proxy and we don't know) so I'm very happy with this as it's the most conservative.

File: tests/erc721/erc721_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (aschlosberg) [reply tests/erc721/erc721_test.go:317]: Makes sense. We can just create a new SimulatedBackend and use the addresses, instead of deploying the entire contract. I've noted it down and will submit changes after I merge this.

**Review by aschlosberg — APPROVED** [review]
This PR is really elegant, I'm loving this review. So many "ahh neat" moments! I really like how you've incorporated the exists and notExists IDs to. It reads like a story
